x-pack/legacy/plugins/apm/public/components/app/TransactionDetails/WaterfallWithSummmary/WaterfallContainer/Waterfall/waterfall_helpers/waterfall_helpers.ts

```diff
@@ -216,79 +213,98 @@ function getDuration(items: IWaterfallItem[]) {
+    waterfallItem =>
+      waterfallItem.docType === 'transaction' && waterfallItem.id === id
+  ) as IWaterfallItemTransaction;
+
+const sortWaterfallWithoutParent = (
+  waterfallItems: Array<IWaterfallItemSpan | IWaterfallItemTransaction>
+) => {
+  const items = sortBy(waterfallItems, 'start');
+
+  items.forEach(item => {
+    item.offset = item.start || 0;
+  });
+  return items;
+};
 
 export function getWaterfall(
   { trace, errorsPerTransaction }: TraceAPIResponse,
   entryTransactionId?: Transaction['transaction']['id']
 ): IWaterfall {
   if (isEmpty(trace.items) || !entryTransactionId) {
     return {
-      services: [],
+      serviceColors: {},
       duration: 0,
-      orderedItems: [],
-      itemsById: {},
-      getTransactionById: () => undefined,
-      errorCountByTransactionId: errorsPerTransaction,
-      serviceColors: {}
+      errorsPerTransaction,
+      items: [],
+      getTransactionById: () => undefined
     };
   }
 
-  const waterfallItems = trace.items.map(traceItem => {
-    const docType = traceItem.processor.event;
-    switch (docType) {
-      case 'span':
-        return getSpanItem(traceItem as Span);
-      case 'transaction':
-        return getTransactionItem(
-          traceItem as Transaction,
-          errorsPerTransaction
-        );
-    }
-  });
+  const waterfallItems = transformTraceItems({ trace, errorsPerTransaction });
 
-  const childrenByParentId = groupBy(waterfallItems, item =>
-    item.parentId ? item.parentId : 'root'
+  const entryWaterfallTransaction = findWaterfallTransactionById(
+    waterfallItems,
+    entryTransactionId
   );
-  const entryTransactionItem = waterfallItems.find(
-    waterfallItem =>
-      waterfallItem.docType === 'transaction' &&
-      waterfallItem.id === entryTransactionId
+
+  const isParentMissing = waterfallItems.every(waterfallItem =>
+    isEmpty(waterfallItem.parentId)
   );
-  const itemsById: IWaterfallIndex = indexBy(waterfallItems, 'id');
-  const orderedItems = entryTransactionItem
-    ? getOrderedWaterfallItems(childrenByParentId, entryTransactionItem)
-    : [];
-  const traceRoot = getTraceRoot(childrenByParentId);
-  const duration = getDuration(orderedItems);
-  const traceRootDuration = traceRoot && traceRoot.transaction.duration.us;
-  const services = getServices(orderedItems);
-  const getTransactionById = createGetTransactionById(itemsById);
-  const serviceColors = getServiceColors(services);
-  const entryTransaction = getTransactionById(entryTransactionId);
+
+  let items;
+  let traceRootTransaction;
+
+  if (isParentMissing) {
+    items = sortWaterfallWithoutParent(waterfallItems);
+  } else {
+    const childrenByParentId = groupBy(waterfallItems, item =>
+      item.parentId ? item.parentId : 'root'
+    );
+
+    items = entryWaterfallTransaction
+      ? sortWaterfall(childrenByParentId, entryWaterfallTransaction)
+      : [];
+
+    traceRootTransaction = getTraceRootTransaction(childrenByParentId);
+  }
```

Comment:
I think this'll be simpler if it is reduced to a ternary. Additionally I think the sort methods can be named more accurately.
```ts
  const items = isParentMissing
    ? sortWaterfallByStartTime(waterfallItems)
    : sortWaterfallByParentId(childrenByParentId, entryWaterfallTransaction);
```